File: ricker/src/matlab/persdia_sub_super.m
function n = persdia_sub_super(varargin)
% n = number of persistence intervals
% inputs: mandatory filename = name of file containing 
% birth and death info in column form, and a number plot-type.
% if plot type = 0, we plot ALL intervals, otherwise only the
% ones that actually die.

if nargin==0
    n = -1;
    return
else
    filename = varargin{1};
end    

plot_type = 0;
if nargin == 2 
    plot_type = varargin{2};
end

% set a flag in case the only death is infinite; default is false
maxd_is_inf = 0; 
normal_gens = 0;

% extract birth and death indices
ints = load(filename);
births = ints(:,1)
deaths = ints(:,2)

% extract indices of those intervals which persist throughout
sub_ints = find(deaths == -1);
super_ints = find(deaths == -2);

% extract indices of those intervals which die
try
    normal_ints = find(deaths ~= -1 & deaths ~= -2);
    normal_gens = 1;
catch
    disp( 'Only infinite generators on this homology level.' );
end

% extract maximum death time encountered and min birth times
minb = min( births ) % should always be >= 1
maxbirth = max( births )
maxd = max( deaths )
if (maxd < 0) 
    maxd_is_inf = 1;
    maxd = max( births )
end

% start figure
figure;

% we always plot these
plot(births(normal_ints),deaths(normal_ints),'bo','MarkerFaceColor','b', ...
     'MarkerSize', 2 );
hold on;

max_axis = max( [ maxd, maxbirth ] )
axis([-0.1*maxd,(1.1)*(max_axis+1),-0.1*maxd, (1.1)*(max_axis+1)]);

% plot the diagonal
%diag = -0.1*maxd : (1.1)*maxd;

title(filename);
xlabel('birth');
ylabel('death');

%% set super-level infinite gens to min - 1 (so if min val in
%% matrix is 2, set inf to 1 )
infsub_vec = (max_axis + 1) * ones(size(sub_ints));
%infsuper_vec = (minb -1 ) * ones(size(super_ints));
infsuper_vec = zeros(size(super_ints));
diag = 0:(1.1)*(max_axis+2)/20:(1.1)*(max_axis+2);
%diag = -0.1*maxd : (1.1)*(maxd+1) 
npts = length( diag );
 
% plot diag and horizontal lines
plot(diag,diag,'g-');
plot( zeros( npts ), diag, 'k--' ); % horizontal line
plot( diag, zeros( npts ), 'k--' ); % vertical line

% plot infinite stuff
plot((births(sub_ints)), infsub_vec , 'rd', 'LineWidth', 1 ); %,'MarkerFaceColor','r' );
plot((births(super_ints)), infsuper_vec , 'rd'); %,'MarkerFaceColor','r');    
    
n = size(births);

hold off;

File: persdia_sub_super.py
import numpy as np
import matplotlib.pyplot as plt


def persdia_sub_super( filename, **kwargs ):
    """
    filename -- path to .npy or .txt file containing persistence intervals

    birth and death info in column form
    
    """
    args = { 'ms' : 2 }
    args.update( kwargs )

    
    # set a flag in case the only death is infinite; default is false
    maxd_is_inf = False 
    normal_gens = False

    # extract birth and death indices
    try:
        ints = np.load( filename )
    except IOError:
        ints = np.loadtxt( filename )
    births = ints[:,0]
    deaths = ints[:,1]

    # extract indices of those intervals which persist throughout
    sub_ints = np.where( deaths == -1 )[0]
    super_ints = np.where( deaths == -2 )[0]

    # extract indices of those intervals which die
    try:
        n1 = np.where( deaths != -1 )[0]
        n2 = np.where( deaths != -2 )[0]
        normal_ints = np.intersect1d( n1, n2 )
        normal_gens = True  # we have normal (non-infinite gens)
    except:
        normal_ints = []
        print 'Only infinite generators on this homology level.' 

    # extract maximum death time encountered and min birth times
    #minb = births.min()  #min( births ) % should always be >= 1
    maxbirth = births.max() #max( births )
    maxd = deaths.max() # max( deaths )
    if maxd < 0:
        maxd_is_inf = True;
        maxd = births.max() #max( births )

    # start drawing the figure
    fig = plt.figure()
    ax = fig.gca()
    
    # we always plot these
    ax.plot( births[ normal_ints ], deaths[ normal_ints ], 'bo', **args  );

    # create diagonal
    diag = [0, maxd+2]
    ax.plot(diag, diag, 'g-')

    # set super-level infinite gens to min - 1 (so if min val in
    # matrix is 2, set inf to 1 )
    max_axis = max( [ maxd, maxbirth ] )
    infsub_vec = (max_axis + 1) * np.ones_like( sub_ints )
    #%infsuper_vec = (minb -1 ) * ones(size(super_ints));
    infsuper_vec = np.zeros_like( super_ints )

    # create diag, from equally spaced points (20 steps)
    diag = np.linspace( 0, (1.1)*(max_axis+2), (1.1)*(max_axis+2)/20 )
    npts = len( diag );

    # plot diag and horizontal lines
    ax.plot(diag,diag,'g-');
    ax.plot( npts * [0], diag, 'k--' )
    ax.plot( diag, npts * [0], 'k--' )

    # plot infinite stuff
    ax.plot( births[ sub_ints ], infsub_vec , 'rd', lw=1, **args )
    ax.plot( births[ super_ints ], infsuper_vec , 'rd', **args )

    # axis lines, dashed
    ax.hlines( 0, 0, max_axis+1, linestyle='dashed' )
    ax.vlines( 0, 0, max_axis+1, linestyle='dashed' )

    # set axis limits
    ax.set_xlim( [ -0.1*maxd, (1.1)*(max_axis+1) ] )
    ax.set_ylim( [ -0.1*maxd, (1.1)*(max_axis+1) ] )

    ax.set_title(filename)
    ax.set_xlabel('birth')
    ax.set_ylabel('death')

    return fig
